### PR TITLE
Add syntax highlighting

### LINKED ISSUE #9: Add syntax highlighting
Implement syntax highlighting from the plugin.

This should re-implement the functionality offered by the TextMate bundle: https://github.com/awslabs/smithy-vscode

### PR DESCRIPTION
**Pull request by srchase**
This PR closes #9, adding syntax highlighting for the Smithy IDL, via `Flex` and `BNF` files used to generate the Lexer and Parser classes.

Two Gradle tasks have been added. `generateSmithyParser` uses the `Smithy.bnf` file to generate the parser classes to `/src/gen/main`. `generateSmithyLexer` uses the `Smithy.flex` file to generate the lexer class to the same path.

During a standard build, using JDK11 or later, the lexer and parser classes will be regenerated prior to compilation. The parser generator requires JDK11. When using an older version of the JDK, these classes will not be re-generated as part f a standard build.

For development, the `Grammar-Kit` plugin is especially useful, adding context menu options to regenerate the Lexer and Parser classes, along with a live-preview for tuning the BNF. However, the `Grammar-Kit` plugin generates slightly different generated classes due to the differences in the vended generators. As a result, only the Gradle tasks should be used for generating classes that get committed to this repository.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/gen/software/amazon/smithy/plugin/language/SmithyLexer.java [hunk 1]
```diff
@@ -0,0 +1,1005 @@
...
+
+import static com.intellij.psi.TokenType.BAD_CHARACTER;
+import static com.intellij.psi.TokenType.WHITE_SPACE;
+import static software.amazon.smithy.plugin.language.psi.SmithyTypes.*;
+
+/**
+ * This class is a scanner generated by JFlex
+ * from the specification file <tt>Smithy.flex</tt>
+ */
+
```
Comment (srchase) [inline src/main/gen/software/amazon/smithy/plugin/language/SmithyLexer.java:31]: There's an issue with the Lexer generator vended in the Grammar-Kit plugin that results in this extra space after the doc.

File: src/main/java/software/amazon/smithy/plugin/language/Smithy.bnf [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+node_array ::= LBRACK ws (node_value ws)* RBRACK
+
+node_object ::= LBRACE ws (node_object_kvp ws)* RBRACE
+
+node_object_kvp ::= IDENTIFIER ws COLON ws node_value
+
+number ::= [MINUS] INT [FRAC] [EXP]
+
+//node_string_value ::= shape_id | TEXT_BLOCK | quoted_text
```
Comment (kstich) [inline src/main/java/software/amazon/smithy/plugin/language/Smithy.bnf:92]: It looks like there are a bunch of leftover productions in comments to clean up.

File: src/main/java/software/amazon/smithy/plugin/language/Smithy.bnf [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    NODE_OBJECT_TYPE_NAME = "regexp:(service|resource|operation)"
+    SHAPE_ID_MEMBER = "regexp:\$(_|\p{Alpha})(_|\p{Alpha}|\p{Digit})*"
+    IDENTIFIER = "regexp:(_|\p{Alpha})(_|\p{Alpha}|\p{Digit})*"
+    TEXT_BLOCK = "regexp:\"\"\"(\r?\n[ !#-\[\]-}]*)*\"\"\""
+  ]
+}
+
+idl ::= ws version_section ws metadata_section shape_section
+
+version_section ::= VERSION COLON ws QUOTED_TEXT
```
Comment (kstich) [inline src/main/java/software/amazon/smithy/plugin/language/Smithy.bnf:72]: If it's not feasible to make this a full `control_section` at the moment, let's leave a note here.

File: src/main/java/software/amazon/smithy/plugin/language/Smithy.bnf [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+trait_body ::= LPAREN ws trait_body_value ws RPAREN
+
+trait_body_value ::= trait_structure | node_value
+
+trait_structure ::= trait_structure_kvp (ws trait_structure_kvp)*
+
+trait_structure_kvp ::= node_object_key ws COLON ws node_value
+
+node_object_key ::= QUOTED_TEXT | IDENTIFIER
```
Comment (kstich) [inline src/main/java/software/amazon/smithy/plugin/language/Smithy.bnf:147]: Can this be up by the other `node_object` pieces?

**Review by kstich — CHANGES_REQUESTED** [review]

File: build.gradle.kts [hunk 5]
```diff
@@ -72,6 +95,36 @@ tasks["checkstyleTest"].enabled = false
...
+ *
+ * Apply Grammar-Kit plugin and configure tasks for generating Lexer and Parser classes.
+ */
+apply(plugin = "org.jetbrains.grammarkit")
+
+grammarKit {
+    jflexRelease = "1.7.0-1"
+}
+
+val generateSmithyLexer = task<GenerateLexer>("generateSmithyLexer") {
```
[the remark this answers is not in the document]
Comment (srchase) [reply build.gradle.kts:114]: Added in ee04ca9 before compiliation.

File: src/main/java/software/amazon/smithy/plugin/language/SmithySyntaxHighlighter.java [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.plugin.language;
```
[the remark this answers is not in the document]
Comment (srchase) [reply src/main/java/software/amazon/smithy/plugin/language/SmithySyntaxHighlighter.java:16]: No, it is not.

File: src/main/java/software/amazon/smithy/plugin/language/SmithySyntaxHighlighter.java [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+            createTextAttributesKey("SMITHY_USE_KEYWORD", DefaultLanguageHighlighterColors.METADATA);
+    public static final TextAttributesKey VERSION =
+            createTextAttributesKey("SMITHY_VERSION", DefaultLanguageHighlighterColors.METADATA);
+    public static final TextAttributesKey VERSION_SECTION =
+            createTextAttributesKey("SMITHY_VERSION_SECTION", DefaultLanguageHighlighterColors.METADATA);
+    public static final TextAttributesKey WS =
+            createTextAttributesKey("SMITHY_WS", DefaultLanguageHighlighterColors.PARAMETER);
+
+    private static final TextAttributesKey[] APPLY_KEYS = new TextAttributesKey[]{APPLY};
+    private static final TextAttributesKey[] BAD_CHAR_KEYS = new TextAttributesKey[]{BAD_CHARACTER};
```
Comment (kstich) [inline src/main/java/software/amazon/smithy/plugin/language/SmithySyntaxHighlighter.java:125]: Nit: `BAD_CHAR_KEYS` is the only shortened name for these, can just keep it similar.

**Review by kstich — APPROVED** [review]

**Pull request merged**